File: 871/solution1.c
// 0/1 knapsack problem

int max(const int a, const int b) {
    return a > b ? a : b;
}

int minRefuelStops(int target, int startFuel, int** stations, int stationsSize, int* stationsColSize){
    int n = stationsSize;
    long *dp = (long *)malloc(sizeof(long) * (n + 1));
    for(int i = 0; i <= n; i++) {
        dp[i] = startFuel;
    }
    
    for(int i = 0; i < n; i++) {
        for(int j = i + 1; j >= 1; j--) {
            if(dp[j - 1] >= stations[i][0]) {
                dp[j] = max(dp[j], dp[j - 1] + stations[i][1]);
            }
        }
    }
    
    for(int i = 0; i <= n; i++) {
        if(dp[i] >= target) {
            return i;
        }
    }
    
    return -1;
}

File: 871/solution1.rb
# 0/1 knapsack problem

# @param {Integer} target
# @param {Integer} start_fuel
# @param {Integer[][]} stations
# @return {Integer}
def min_refuel_stops(target, start_fuel, stations)
    n = stations.size
    dp = Array.new(n + 1, start_fuel)
    
    0.upto(n - 1) do |i|
        (i + 1).downto(1) do |j|
            dp[j] = [dp[j], dp[j - 1] + stations[i][1]].max if dp[j - 1] >= stations[i][0]
        end
    end
    
    dp.size.times do |i|
        return i if dp[i] >= target
    end
    
    -1
end
